<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>promise4</title>
</head>

<body>
    <script>
        const wait = (ms, ok) => {
            return new Promise((resolve, reject) => {
                setTimeout(() => {
                    if (ok) {
                        resolve('ok: ' + ms)
                    } else {
                        reject(new Error('no ok:' + ms))
                    }
                }, ms)
            })
        }

        // Promise.all(Promise陣列)
        // 同時開始執行Promise
        /* 陣列裡的所有東西都resolve才會.then()
        只要有一個reject就會catch */
        Promise.all([
            wait(1000, true),
            wait(2000, false),
            wait(3000, true)
        ])
            .then(result => {
                console.log(result)
            })
            .catch(error => {
                console.log(error)
            })

        // Promise.allSettled(Promise陣列)
        // 同時開始執行Promise
        /* 陣列裡的所有東西都處理完後直接.then
        .then的status代表結果
        fulfilled成功 、 rejected失敗
        value 表示成功產生的結果、reason 表示失敗的原因
         */
        Promise.allSettled([
            wait(1000, true),
            wait(2000, false),
            wait(3000, true)
        ])
            .then(result => {
                console.log(result)
            })

        // Promise.race(Promise陣列)
        // 取陣列內第一個有結果的東西
        Promise.race([
            wait(1000, true),
            wait(2000, false),
            wait(3000, true)
        ])
            .then(result => {
                console.log(result)
            })
            .catch(error => {
                console.log(error)
            })

        // Promise.any(Promise陣列)
        // 取陣列內第一個resolve的東西
        // 全部reject會進到catch，錯誤是all promise were rejected
        Promise.any([
            wait(1000, false),
            wait(2000, false),
            wait(3000, true)
        ])
            .then(result => {
                console.log(result)
            })
            .catch(error => {
                console.log(error)
            })


    </script>
</body>

</html>